### ISSUE TITLE
Phase 3: Auth pipeline MVP (app-first) + liveness

### ISSUE THREAD

**Issue by CipherxHub** [CipherxHub/SecureFaceUnlock#1]
# Phase 3: Auth pipeline MVP (app-first) + liveness

## Objective

Build the **authentication decision pipeline** as a runnable app (not yet a Windows
service): capture a face, verify it is live, match it against an enrolled template,
and return a **GRANT/DENY** verdict. This is the component Phase 4 will wrap in a
named-pipe service.

## Context (reuse these, do not re-implement)

| Module | Path | What it provides |
|---|---|---|
| Detection | `face_engine/detector/yunet.py` | `YuNetDetector.detect(frame) -> list[FaceDetection]` (bbox, score, 5 landmarks) |
| Quality gate | `face_engine/quality/quality_checker.py` | `FaceQualityChecker.from_config(cfg)`, `check(frame, faces) -> (ok, reason)` |
| Alignment | `face_engine/alignment/aligner.py` | `align_face(frame, landmarks) -> 112x112 BGR crop` |
| Embedding | `face_engine/recognizer/sface.py` | `SFaceEmbedder.embed(aligned_crop) -> 128-d L2-normalized f32` |
| Storage | `enrollment/store.py` | `TemplateStore.from_config(cfg)`, `.load(username) -> FaceTemplate`, `.list()` |
| CLI pattern | `enrollment/enroll.py` | argparse subcommand structure, `_load_config()`, `_components()` |
| Config | `config/config.yaml` | `face_recognition`, `liveness`, `authentication`, `quality` sections |

## Deliverables

1. `face_engine/pipeline/auth.py` — the decision engine:
   - `AuthConfig` dataclass + `from_config(d)` (reads `face_recognition` + `liveness` + `authentication`).
   - `authenticate(frame, username) -> AuthResult` where `AuthResult` carries
     `granted: bool`, `similarity: float`, `liveness_score: float`, `reason: str`.
   - Order: **detect -> quality gate (reject immediately) -> align -> embed -> compare ->
     liveness -> decision**. `granted` requires BOTH `similarity >= threshold` AND `liveness == real`.
   - **Multi-frame voting**: per `config.yaml`, require `multi_frame_required` matches out of
     `multi_frame_total` consecutive frames before granting (reduces false accepts).
2. `face_engine/liveness/fasnet.py` — MiniFASNet wrapper:
   - Model facts (verified): ONNX input `1,3,80,80` float, output `1,3` logits.
   - `FASNetDetector` mirrors the `SFaceEmbedder` style; softmax the logits, take the
     **real-class** score (MiniFASNet v2 classes: print-fake, replay-fake, real — confirm the
     index empirically against a webcam/photo before finalizing).
   - Temporal check: require a live score above `score_threshold` (0.5) across
     `temporal_window_frames` (15) as configured.
3. `scripts/auth_demo.py` — CLI: `python scripts/auth_demo.py <username>` opens the webcam,
   streams the pipeline verdict live (mirror `detect_demo.py` HUD), exits 0 on unlock (you may
   print "UNLOCKED") / 1 on DENY. Accept `--image` for a single-shot test.
4. Tests in `tests/`:
   - `test_fasnet.py` — shape/softmax/score range on the real model (skip if model missing,
     same pattern as `test_sface.py`).
   - `test_auth.py` — synthetic embeddings: granted when sim >= threshold + live; denied when
     sim below threshold; denied when liveness fake; quality-gate refusal of blank frames.
   - `test_auth_demo.py` — single-shot `--image` against a blank frame returns nonzero.

## Important decisions / gotchas

- **Threshold is calibrated, do not lower it.** `similarity_threshold: 0.50` was measured on
  the real model (impostor pairs ~0.42, genuine variation ~0.75–0.95). Re-verify with your own
  camera but keep 0.50 as default.
- **Liveness order:** run it last, only after a template match candidate exists (cheap win,
  fewer model passes).
- **Never log or persist** a raw frame, the embedding outside the store, or any password.
- Webcam garbage frames that produce NaN embeddings must be treated as DENY, never crash the loop.

## Definition of Done (PR checklist)

- [ ] `face_engine/pipeline/auth.py` + `face_engine/liveness/fasnet.py` implement config-driven pipeline.
- [ ] `scripts/auth_demo.py` streams verdicts; `--image` mode returns 0/1.
- [ ] All new tests pass locally (`python -m pytest`) and on CI.
- [ ] `config.yaml` unchanged except where the pipeline genuinely needs a new key (justify it).
- [ ] Follows repo conventions (dataclasses, `from __future__ import annotations`, `from_config`).
- [ ] Red-team constraints hold: no plaintext secrets, quality gate before embedding, DENY on error.

## Verification

```powershell
python scripts/download_models.py
python scripts/verify_models.py
python -m pytest tests/test_fasnet.py tests/test_auth.py tests/test_auth_demo.py -v
python scripts/auth_demo.py <your-user>          # paired with an enrolled template
python scripts/auth_demo.py --image anything.jpg # must refuse garbage
```

## Working together (parallel-safe)

- Branch: `phase/3-auth-pipeline` off `main`.
- Keep changes scoped to the deliverables; do **not** refactor Phase 1–2 code unless bug-for-bug.
- Rebase on `main` before opening the PR (`git fetch origin && git rebase origin/main`).
- One PR, title `Phase 3: auth pipeline + liveness`, body `Closes #<this issue>`, reference your fixes.
- The merge strategy is squash-merge onto `main` so parallel phase branches stay rebaseable;
  teammates working on Phase 4/5 must rebase onto the same `main`.

**Comment by CipherxHub** [comment]
Phase 3 complete and verified.

- \ace_engine/pipeline/auth.py\: detect -> quality -> align -> embed -> compare -> liveness -> multi-frame vote (\multi_frame_required\ of \multi_frame_total\), GRANT/DENY with reasons, NaN guards, never crashes.
- \ace_engine/liveness/fasnet.py\: MiniFASNet-V2 wrapper (2.7x margin crop, BGR /255 NCHW, live = softmax[0]).
- \scripts/auth_demo.py\: webcam streaming + \--image\ 0/1, \--headless\ for CI.
- Hardening: \landmarks_usable()\ + bbox-crop fallback in aligner (FaceDetectorYN returns keypoints in wrong coords on some frames); quality tilt gate skips unusable landmarks.
- 15 new tests, 51 total; CI green (python-tests + cpp-build).

Verified end-to-end: photo enrollment succeeds, then the SAME photo is DENIED at liveness (score 0.000, classified 'replay') — presentation attack blocked.

Note: temporal liveness is enforced per-frame inside the multi-frame vote window rather than a separate 15-frame rolling window; \	emporal_window_frames\ remains as the configured cap.
